src/Open3D/Visualization/Rendering/Filament/TriangleMeshBuffers.cpp

```diff
@@ -0,0 +1,469 @@
+        float4VTangents = static_cast<math::quatf*>(malloc(tangentsBytesCount));
+        auto orientation = filament::geometry::SurfaceOrientation::Builder()
+                                   .vertexCount(nVertices)
+                                   .normals(reinterpret_cast<math::float3*>(
+                                           normals.data()))
+                                   .build();
+        orientation.getQuats(float4VTangents, nVertices);
+    } else {
+        utility::LogWarning(
+                "Trying to create mesh without vertex normals. Shading would "
+                "not work correctly. Consider to generate vertex normals "
+                "first.");
+    }
+
+    // We defaulting to have vertex color attribute for all geometries, even if
+    // a geometry doesn't have one That's all due to our default material and
+    // large variety of geometries it should support
+    const bool hasColors = true;  // geometry_.HasVertexColors();
+    const bool hasUVs = geometry_.HasTriangleUvs();
+
+    std::tuple<vbdata, ibdata> buffersData;
+    if (hasUVs) {
+        buffersData = CreateTexturedBuffers(float4VTangents, geometry_);
+    } else if (hasColors) {
+        buffersData = CreateColoredBuffers(float4VTangents, geometry_);
+    } else {
+        buffersData = CreatePlainBuffers(float4VTangents, geometry_);
+    }
+
+    free(float4VTangents);
+
+    const vbdata& vertexData = std::get<0>(buffersData);
+    const ibdata& indexData = std::get<1>(buffersData);
+
+    size_t stride = sizeof(BaseVertex);
+    VertexBuffer* vbuf = nullptr;
+    if (hasUVs) {
+        stride = sizeof(TexturedVertex);
+    } else if (hasColors) {
+        stride = sizeof(ColoredVertex);
+    }
+
+    vbuf = BuildFilamentVertexBuffer(engine, vertexData.verticesCount, stride,
+                                     hasUVs, hasColors);
+
+    VertexBufferHandle vbHandle;
+    if (vbuf) {
+        vbHandle = resourceManager.AddVertexBuffer(vbuf);
+    } else {
+        free(vertexData.bytes);
+        free(indexData.bytes);
+
+        return {};
+    }
+
+    VertexBuffer::BufferDescriptor vertexbufferDescriptor(
+            vertexData.bytes, vertexData.bytesToCopy);
+    vertexbufferDescriptor.setCallback(
+            GeometryBuffersBuilder::DeallocateBuffer);
+    vbuf->setBufferAt(engine, 0, std::move(vertexbufferDescriptor));
+
+    auto ibHandle = resourceManager.CreateIndexBuffer(
+            indexData.bytesCount / indexData.stride, indexData.stride);
+    auto ibuf = resourceManager.GetIndexBuffer(ibHandle).lock();
+
+    // Moving copied indices to IndexBuffer
+    // they will be freed later with freeBufferDescriptor
+    IndexBuffer::BufferDescriptor indicesDescriptor(indexData.bytes,
+                                                    indexData.bytesCount);
+    indicesDescriptor.setCallback(GeometryBuffersBuilder::DeallocateBuffer);
+    ibuf->setBuffer(engine, std::move(indicesDescriptor));
```

Comment:
Presumably not, since we just created it. But if we do, or want to be bullet-proof in our out-of-memory handling, presumably we need to do it for the others, too.